### PR TITLE
Fix dcc ident and format truncation warning

### PR DESCRIPTION
**Pull request by michaelortmann**
Found by: Geo
Patch by: michaelortmann
Fixes: 

One-line summary:
Fix dcc ident and format truncation warning

Additional description (if needed):
Fixes the following compiler warning introduced with 581b71ea140990cd9c653fe28514f3188be74e55:
```
CFLAGS="-Wformat-truncation" ./configure
[...]
make
[...]
dcc.c: In function ‘dcc_ident’:
dcc.c:2257:36: warning: ‘%s’ directive output may be truncated writing up to 511 bytes into a region of size 345 [-Wformat-truncation=]
 2257 |       snprintf(buf1, sizeof buf1, "%s@%s", uid, dcc[idx].host);
      |                                    ^~      ~~~
dcc.c:2257:7: note: ‘snprintf’ output between 2 and 836 bytes into a destination of size 345
 2257 |       snprintf(buf1, sizeof buf1, "%s@%s", uid, dcc[idx].host);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Additionally:
- the following ident truncation to 20 chars was upped to 512 chars:
https://github.com/eggheads/eggdrop/blob/38d675016c87195fdc7c935b93ba7bca3c44afac/src/dcc.c#L2253
If i remember correctly, we got reports of dcc ident not working sometimes, and we never found the reason.
This could be the fix.
- treat an at-char in ident as a failed ident
- the loop was enhanced to not copy a static string over and over

Test cases demonstrating functionality (if applicable):
Now i need someone to test the new dcc ident code. Pls step in someone.

### CODE REVIEW

File: src/dcc.c [hunk 1]
```diff
@@ -2239,24 +2239,30 @@ struct dcc_table DCC_IDENTWAIT = {
...
-  uid[20] = 0;                  /* 20 character ident max */
-  for (i = 0; i < dcc_total; i++)
-    if ((dcc[i].type == &DCC_IDENTWAIT) &&
-        (dcc[i].sock == dcc[idx].u.ident_sock)) {
-      snprintf(buf1, sizeof buf1, "%s@%s", uid, dcc[idx].host);
-      dcc_telnet_got_ident(i, buf1);
-    }
+  rmspace(buf1);
+  buf1[(sizeof buf1) - 1] = 0;
+  if (!strchr(buf1, '@')) {
```
Comment (vanosg) [inline src/dcc.c]: Let's add a debug message saying we rejected the ident, so it doesn't silently reject

File: src/dcc.c [hunk 2]
```diff
@@ -2239,24 +2239,30 @@ struct dcc_table DCC_IDENTWAIT = {
...
+    c = buf1 + strlen(buf1);
+    *c++ = '@';
```
Comment (vanosg) [inline src/dcc.c]: if buf1 is a full 512 characters, is this going to overflow? should buf1 be 513?

File: src/dcc.c (hunk 1 above)
Comment (michaelortmann) [reply src/dcc.c]: Done. It will report "bogus ident response" with log level DEBUG.

File: src/dcc.c (hunk 2 above)
Comment (michaelortmann) [reply src/dcc.c]: I modified sscanf to read up to 511 chars. lets keep buffer sizes to powers of 2.

**Comment by michaelortmann** [comment]
Enhanced this fix, better parser and better logging for identing incomming dcc / telnet.

**Comment by michaelortmann** [comment]
Developer test: set ident-timeout to 5, so that the bot idents incomming users.
After connecting via telnet to the bot and if ident was successful, it looks like:
```
[04:39:13] Telnet connection: localhost.home.arpa/43768
[04:39:17] pbkdf2 method SHA256 rounds 16000, user 3.389ms sys 0.000ms
[04:39:17] Logged in: testuser (michael@localhost.home.arpa/43768)
*** testuser joined the party line.
[04:39:18] BotA joined #test6888.
.who
[04:39:19] tcl: builtin dcc call: *dcc:who -HQ 1 
[04:39:19] #-HQ# who
Party line members: (* = owner, + = master, % = botmaster, @ = op, ^ = halfop)
  [01]  *-HQ       llama@console (con:mkcoblxsd)
  [09]  *testuser  michael@localhost.home.arpa (con:mkcoblxsd)
```
If ident was not successful, like with bogus ident string incomming, it looks like:
```
[04:06:48] Telnet connection: localhost.home.arpa/42056
[04:06:48] dcc: dcc_ident(): ident response port_on_server 5588 doesnt match dcc[4].port 42056
[04:06:48] net: eof!(read) socket 12
[04:06:48] EOF ident connection
[04:06:51] pbkdf2 method SHA256 rounds 16000, user 4.930ms sys 0.000ms
[04:06:51] Logged in: testuser (telnet@localhost.home.arpa/42056)
*** testuser joined the party line.
.[04:06:53] BotA joined #test6888.
who
[04:06:54] tcl: builtin dcc call: *dcc:who -HQ 1 
[04:06:54] #-HQ# who
Party line members: (* = owner, + = master, % = botmaster, @ = op, ^ = halfop)
  [01]  *-HQ       llama@console (con:mkcoblxsd)
  [09]  *testuser  telnet@localhost.home.arpa (con:mkcoblxsd)
```

File: src/dcc.c [hunk 3]
```diff
@@ -2239,27 +2239,40 @@ struct dcc_table DCC_IDENTWAIT = {
...
+  size_t l;
   int i;
 
-  *response = *uid = '\0';
-  sscanf(buf, "%*[^:]:%[^:]:%*[^:]:%[^\n]\n", response, uid);
-  rmspace(response);
-  if (response[0] != 'U') {
-    dcc[idx].timeval = now;
+  *resp_type = *add_info = *user_id = '\0'; 
+  if ((sscanf(buf, "%u ,%u : %511[^ :] : %511[^ :] :%511s", &port_on_server,
```
Comment (thommey) [inline src/dcc.c]:
```
Notes on Syntax:

   1)   To promote interoperability among variant
        implementations, with respect to white space the above
        syntax is understood to embody the "be conservative in
        what you send and be liberal in what you accept"
        philosophy.  Clients and servers should not generate
        unnecessary white space (space and tab characters) but
        should accept white space anywhere except within a
        token.  In parsing responses, white space may occur
        anywhere, except within a token.
```
Coming from an IRC background it is very important to follow the "be liberal in what you accept" rule with so many ircds out there, so we should also do this for ident, which means not requiring those spaces to be there

Comment (michaelortmann) [reply src/dcc.c]: exactly. thats what i did read and implement and developer testet with 0 whitespaces and with 3 consecutive whitespaces.
```
White space (such as
     blanks, tabs, or newlines)	in the format string match any amount of white
     space, including none, in the input.
```
source: https://www.freebsd.org/cgi/man.cgi?query=sscanf&apropos=0&sektion=0&manpath=FreeBSD+13.1-RELEASE+and+Ports&arch=default&format=html

**Comment by michaelortmann** [comment]
The patch has been rewritten to be more simple, so skip the last comments and look only at the new code.ready for review again.

**Pull request merged**